### PR TITLE
[openssl] remove run-time dependency of perl due to c_rehash

### LINKED ISSUE #19965: openssl.bin depends on perl because of c_rehash
Trying to build a small docker image of a nodejs project I noticed that `openssl` depends on `perl`:

```
3.4M    /nix/store/55azyw1bcrzn8q5ganaav0cnqs2viwdn-openssl-1.0.2i
48M     /nix/store/xcp6lrq3ynj75cqcyy7b32jgizf9jr4b-perl-5.22.2
```

The reason for this runtime dependency is the inclusion of `c_rehash`:

```
# Perl c_rehash script, scan all files in a directory
# and add symbolic links to their hash values.
```

There are no references to `c_rehash` in nixpkgs. Would any oppose dropping this from `openssl-bin`?

/cc @edolstra

### PR DESCRIPTION
**Pull request by adrian-gierakowski**
###### Description of changes

Replaces perl based c_rehash script with shell script wrapping `openssl rehash`
with the same functionality.

Fixes: #19965
Supersedes: #156776, #83446
Possibly related to: #157093, #82924

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adrian-gierakowski** [comment]
as discussed [here](https://github.com/NixOS/nixpkgs/pull/156776#issuecomment-1453198061) the `withPerl` flag could be removed altogether

happy to do this if there are no objections

**Comment by Izorkin** [comment]
Need to add these changes for `quictls`?

**Comment by adrian-gierakowski** [comment]
> Need to add these changes for `quictls`?

Was just going to ask 😄

Once the final decision is made regarding  keeping\removing the withPerl flag I’ll apply the changes there as well

**Comment by adrian-gierakowski** [comment]
@veprbl I saw you did code review on #156776 

Do you have the power to merge this? What's your opinion regarding removing the `withPerl` flag? Thanks!

File: pkgs/development/libraries/openssl/default.nix [hunk 1]
```diff
@@ -184,6 +191,13 @@ let
       # "#!/usr/bin/env perl"
     ''
       substituteInPlace $out/bin/c_rehash --replace ${buildPackages.perl}/bin/perl "/usr/bin/env perl"
+    '' + lib.optionalString (!withPerl)
+    # Replace perl based c_rehash script with c_rehash_shim
+    ''
+      substitute \
+        ${c_rehash_shim} \
+        $out/bin/c_rehash \
+        --replace /usr/bin/openssl "${placeholder "bin"}/bin/openssl"
```
Comment (Artturin) [inline pkgs/development/libraries/openssl/default.nix:200]:
```suggestion
    '' + lib.optionalString (!withPerl) ''
      makeWrapper $bin/bin/openssl $bin/bin/c_rehash \
        --add-flags "rehash"
```

**Comment by Artturin** [comment]
let's drop withPerl

**Comment by adrian-gierakowski** [comment]
@Artturin removed `withPerl` and applied your improvement to the wrapper scripts
also applied the same changes to `quictls`

**Comment by Artturin** [comment]
Needs to target staging and the commit msg's need to be changed to our style, instructions for both in `CONTRIBUTING.md`

**Comment by thomasjm** [comment]
Hey, there's about a month until the 23.05 release and it would be awesome to get this merged before then. Are there any blockers remaining?

**Comment by Artturin** [comment]
@ofborg build openssl

**Comment by Artturin** [comment]
Needs to target staging check CONTRIBUTING.md for how to rebase

**Comment by Artturin** [comment]
oh already said that

**Comment by Artturin** [comment]
Can't push to this PR `ERROR: Permission to rhinofi/nixpkgs.git denied to Artturin.` even though `Maintainers are allowed to edit this pull request.`

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/226108 kept your authorship

**Comment by adrian-gierakowski** [comment]
@Artturin thanks! I didn’t really have time to address your comment so it’s great to took care of it :)

**Pull request closed without merging**